locator/token_metadata.cc

```diff
@@ -535,15 +536,14 @@ class token_metadata_impl final {
 
 #endif
     std::vector<inet_address> get_all_endpoints() const {
-        std::vector<inet_address> tmp;
-        std::transform(_endpoint_to_host_id_map.begin(), _endpoint_to_host_id_map.end(), std::back_inserter(tmp), [](const auto& p) {
-           return p.first;
-        });
-        return tmp;
+        std::unordered_set<inet_address> tmp;
+        boost::copy_range<std::unordered_set<gms::inet_address>>(_token_to_endpoint_map | boost::adaptors::map_values);
+        std::vector<inet_address> ret(tmp.begin(), tmp.end());
+        return ret;
     }
 
     size_t get_all_endpoints_count() const {
-        return _endpoint_to_host_id_map.size();
+        return count_normal_token_owners();
```

Comment:
But `count_normal_token_owners` - as the name indicates - takes into account only owners of NORMAL tokens. But what about bootstrapping nodes, for example? `get_all_endpoints_count` talks about _all_ endpoints, so shouldn't it also count nodes that only own pending tokens etc.?

Reply:
Pending node is talked explicitly.

 E.g., 

storage_proxy::create_write_response_handler_helper

std::vector<gms::inet_address> pending_endpoints = get_token_metadata_ptr()->pending_endpoints_for(token, keyspace_name);

Reply:
We  reply on the pending_endpoints_for for the pending nodes.

Reply:
I don't understand how that answers my question.
My question is about what `get_all_endpoints_count()` _means_.

The name suggests that it means "give me all endpoints present in the token ring".
In other words: "give me all endpoints that have at least one token in the token ring".

But this new definition only counts endpoints which have NORMAL tokens, and ignores the others.

In particular, with this change it's not always true that `get_all_endpoints_count() == get_all_endpoints().size()`, which doesn't make any sense to me.

`get_all_endpoints_count()` should always be equal to `get_all_endpoints().size()`, and the implementation may optimize this, but the result should always be the same.

Reply:
Perhaps what you _really_ wanted to achieve in this PR is to modify `everywhere_replication_strategy::get_replication_factor()` like this:
```
size_t everywhere_replication_strategy::get_replication_factor() const {
    return _shared_token_metadata.get()->count_normal_token_owners();
}
```
?

Reply:
I think get_all_endpoints_count means the number nodes in the ring, excluding pending nodes.

Reply:
> In particular, with this change it's not always true that get_all_endpoints_count() == get_all_endpoints().size(), which doesn't make any sense to me.

Why? Now get_all_endpoints_count returns nodes in the ring. get_all_endpoints returns the nodes in the ring too.

Reply:
```
1649 size_t token_metadata_impl::count_normal_token_owners() const {
1650     std::set<inet_address> eps;
1651     for (auto [t, ep]: _token_to_endpoint_map) {
1652         eps.insert(ep);
1653     }
1654     return eps.size();
1655 }

 538     std::vector<inet_address> get_all_endpoints() const {
 539         std::unordered_set<inet_address> tmp;
 540         boost::copy_range<std::unordered_set<gms::inet_address>>(_token_to_endpoint_map | boost::adaptors::map_values);
 541         std::vector<inet_address> ret(tmp.begin(), tmp.end());
 542         return ret;
 543     }
 544 
 545     size_t get_all_endpoints_count() const {
 546        
```

Reply:
Ok. Now I remember that `_token_to_endpoint_map` contains only normal tokens (sadly it's undocumented so I'm not sure), so `get_all_endpoints()` returns only endpoints which have a normal token after this fix.

Reply:
In any case, `get_all_endpoints` should definitely get a better name and a comment

Reply:
@asias I don't think you resolved this comment (or @kbr- 's concern, which worries me too). It's not clear **exactly** which nodes get_all_endpoints() and get_all_endpoints_count(), list, and that in both of them we actually list exactly the same nodes. 
What really adds to the confusion is that you could have made get_all_endpoints_count() simply return get_all_endpoints().size(), but you decided to call a different pre-existing function instead. Why? Does this function do something different or the same thing? It would be best to remove this duplication and remove one of the two functions, or if you don't want to do that, at least add a comment what that get_all_endpoints_count() does (that it's the same as get_all_endpoints().size()).

Reply:
@asias this function still needs a comment (and/or a rename) to say that it it only counts "normal" (not pending, streaming, nodes). 
You also have not yet addressed my question why, we have both count_normal_token_owners() and the functions you add in this patch. It's code duplication, and if they don't use the same algorithm the better one should be kept and the other removed.

Reply:
I think I understand better now. The new get_all_endpoints() and the old count_normal_token_owners() use exactly the same algorithm, and I would be more comfortable if they were next to each other in the source code to make this clearer. I also don't understand why we need an alias get_all_endpoints_count() to the existing count_normal_token_owners(). But I tend to agree this is mostly cosmetics that can be fixed later - if you remember to do it...

Reply:
Resolving this discussion. I see you're removing this function in the fourth patch. You could have squashed the two patches. But I won't insist.